src/parser/utils.rs

```diff
@@ -100,6 +104,114 @@ impl InfixOp {
     }
 }
 
+/// General interface for structures representing a series of annotations.
+pub trait Annot: Default + From<MetaValue> {
+    /// Attach the annotation to a term.
+    fn attach(self, value: RichTerm, pos: TermPos) -> RichTerm;
+    /// Combine two annotations.
+    fn combine(outer: Self, inner: Self) -> Self;
+}
+
+/// Consistency of naming inside the parser module.
+pub type MetaAnnot = MetaValue;
+
+impl Annot for MetaValue {
+    fn attach(mut self, value: RichTerm, pos: TermPos) -> RichTerm {
+        self.value = Some(value);
+        RichTerm::new(Term::MetaValue(self), pos)
+    }
+
+    fn combine(outer: Self, inner: Self) -> Self {
+        Self::flatten(outer, inner)
+    }
+}
+
+/// Used to combine annotations in a pattern. If at least one annotation is not `None`, then this
+/// just calls [`Annot::combine`] and substitute a potential `None` by the default value.
+///
+/// If both arguments are `None`, we still need a label to report useful error diagnostics. In this
+/// case, `combine_match_annots` returns a value with a dummy `Dyn` contract and a label with the
+/// position set to `span`.
+pub fn combine_match_annots(
+    anns: Option<MetaAnnot>,
+    default: Option<MetaAnnot>,
+    span: RawSpan,
+) -> MetaAnnot {
+    match (anns, default) {
+        (anns @ Some(_), default) | (anns, default @ Some(_)) => {
+            Annot::combine(anns.unwrap_or_default(), default.unwrap_or_default())
+        }
+        (None, None) => MetaValue {
+            contracts: vec![Contract {
+                types: Types(AbsType::Dyn().into()),
+                label: Label {
+                    span,
+                    ..Default::default()
+                },
+            }],
+            ..Default::default()
+        },
+    }
+}
+
+/// Some constructs are introduced with the metavalue pipe operator `|`, but aren't metadata per se
+/// (ex: `push force`/`push default`). Those are collected in this extended annotation and then
+/// desugared into a standard metavalue.
+#[derive(Clone, Debug, Default)]
+pub struct ExtdAnnot {
+    /// Standard metadata.
+    pub meta: MetaValue,
+    /// Presence of an annotation `push force`
+    pub push_force: bool,
+    /// Presence of an annotation `push default`
+    pub push_default: bool,
```

Comment:
Does it ever make sense for a value to be annotated with both `_push_force` and `_push_default`? If not, I'm wondering if it could make more sense to model this in a way where it's not possible for both to be `true` at the same time, e.g.

```rust
pub push_prio: Option<PushPriority>
```

Reply:
It's indeed fishy semantically, but the thing is that nothing currently prevent the user from doing so:

```
x | _push_default | _push_force = ...
```

because annotations are just parsed one by one and combined. In some sense, you can see `ExtdAnnot` as an AST for meta-annotation, and the source syntax does allow both to be set, just as they were flags. One it's actually transformed into a term by `attach()`, it does have to chose an actual `push_prio` to use.